fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/sorting/RecordQuerySortPlan.java

```diff
@@ -0,0 +1,220 @@
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlanWithChild;
+import com.apple.foundationdb.record.query.plan.temp.AliasMap;
+import com.apple.foundationdb.record.query.plan.temp.CorrelationIdentifier;
+import com.apple.foundationdb.record.query.plan.temp.GroupExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.Quantifier;
+import com.apple.foundationdb.record.query.plan.temp.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.temp.explain.NodeInfo;
+import com.apple.foundationdb.record.query.plan.temp.explain.PlannerGraph;
+import com.apple.foundationdb.record.query.predicates.QueriedValue;
+import com.apple.foundationdb.record.query.predicates.Value;
+import com.apple.foundationdb.record.sorting.FileSortCursor;
+import com.apple.foundationdb.record.sorting.MemorySortCursor;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * A query plan implementing sorting in-memory, possibly spilling to disk.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class RecordQuerySortPlan implements RecordQueryPlanWithChild {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Record-Query-Sort-Plan");
+
+    @Nonnull
+    private final Quantifier.Physical inner;
+    @Nonnull
+    private final RecordQuerySortKey key;
+
+    public RecordQuerySortPlan(@Nonnull RecordQueryPlan plan, @Nonnull RecordQuerySortKey key) {
+        this(Quantifier.physical(GroupExpressionRef.of(plan)), key);
+    }
+
+    private RecordQuerySortPlan(@Nonnull Quantifier.Physical inner, @Nonnull RecordQuerySortKey key) {
+        this.inner = inner;
+        this.key = key;
+    }
+
+    @Nonnull
+    @Override
+    public <M extends Message> RecordCursor<QueryResult> executePlan(@Nonnull FDBRecordStoreBase<M> store,
+                                                                     @Nonnull EvaluationContext context,
+                                                                     @Nullable byte[] continuation,
+                                                                     @Nonnull ExecuteProperties executeProperties) {
+        // Since we are sorting, we need to feed through everything from the inner plan,
+        // even just to get the top few.
+        final ExecuteProperties executeInner = executeProperties.clearSkipAndLimit();
+        final Function<byte[], RecordCursor<FDBStoredRecord<M>>> innerCursor =
+                innerContinuation -> getChild().execute(store, context, innerContinuation, executeInner).map(FDBQueriedRecord::getStoredRecord);
```

Comment:
future proofing this may be better to call `executePlan` and `.map(result -> result.getQueriedRecord(0))` as this will eventually have the multi-column results